bmt/bmt_r.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// simple nonconcurrent reference implementation for hashsize segment based
+// Binary Merkle tree hash on arbitrary but fixed maximum chunksize
+//
+// This implementation does not take advantage of any paralellisms and uses
+// far more memory than necessary, but it is easy to see that it is correct.
+// It can be used for generating test cases for optimized implementations.
+// see testBMTHasherCorrectness function in bmt_test.go
+package bmt
+
+import (
+	"hash"
+)
+
+// RefHasher is the non-optimized easy to read reference implementation of BMT
+type RefHasher struct {
+	span    int
+	section int
+	cap     int
+	h       hash.Hash
+}
+
+// NewRefHasher returns a new RefHasher
+func NewRefHasher(hasher BaseHasher, count int) *RefHasher {
+	h := hasher()
+	hashsize := h.Size()
+	maxsize := hashsize * count
+	c := 2
+	for ; c < count; c *= 2 {
+	}
+	if c > 2 {
+		c /= 2
+	}
+	return &RefHasher{
+		section: 2 * hashsize,
+		span:    c * hashsize,
+		cap:     maxsize,
+		h:       h,
+	}
+}
+
+// Hash returns the BMT hash of the byte slice
+// implements the SwarmHash interface
+func (rh *RefHasher) Hash(d []byte) []byte {
+	if len(d) > rh.cap {
+		d = d[:rh.cap]
+	}
+
+	return rh.hash(d, rh.span)
+}
+
+func (rh *RefHasher) hash(d []byte, s int) []byte {
+	l := len(d)
+	left := d
+	var right []byte
+	if l > rh.section {
```

Comment:
Should we not be handling the case where `hashsize < l <= 2*hashsize` here and calculating the hash of both left and right?

For example, consider the BMT of 64 zero bytes. I would expect it to be the sha3 of the concatenation of two sha3 hashes of 32 zero bytes:

```go
sha3 := func(data ...[]byte) []byte {
	h := sha3.NewKeccak256()
	for _, v := range data {
		h.Write(v)
	}
	return h.Sum(nil)
}
zeroes := make([]byte, 32)
fmt.Printf("BMT of 64 bytes: %x\n", sha3(sha3(zeroes), sha3(zeroes)))
```

Running that I get:

```
$ go run bmt.go 
BMT of 64 bytes: 633dc4d7da7256660a892f8f1604a44b5432649cc8ec5cb3ced4c4e6ac94dd1d
```

However this reference implementation gives a different hash (which is just `sha3(zeroes + zeroes)`):

```go
h := bmt.NewRefHasher(sha3.NewKeccak256, 128)
fmt.Printf("Ref BMT of 64 bytes: %x\n", h.Hash(make([]byte, 64)))
```

```
$ go run ref-bmt.go 
Ref BMT of 64 bytes: ad3228b676f7d3cd4284a5443f17f1962b36e491b30a40b2405849e597ba5fb5
```

Am I missing something here?